### ISSUE TITLE
about implementation details in paper

### ISSUE THREAD

**Issue by Cverlpeng** [cardwing/Codes-for-Lane-Detection#145]
hi, 
工作非常棒， 对你的工作非常感兴趣，有一些实现细节请教一下你，但是torch框架不太熟悉。
1.resnet101:你能详细说明一下你的实验中使用的resnet101是下采样了8倍，然后在阶段3和阶段4中使用空洞卷积代替普通卷积吗？还有就是它是否有head部分呢？也就是有没有将最后一层卷积输出的2048的通道数过度到128的部分？类似于SCNN
2.我没太看明白你的SAD中，蒸馏损失函数的输入你是通过求平方，然后在通道上求和使其变成（H*W）的吗？然后，你再使用softmax的时候它的channel不是等于1了吗？请问你的softmax计算是从哪个维度运算的？
期待回复
非常感谢你

**Comment by cardwing** [comment]
1. Yes. We use 8 x downsampling for ResNet-101 and it does not have the message passing module of SCNN.
2. Softmax is performed in the spatial dimension, i.e., HxW.

**Comment by Cverlpeng** [comment]
> 1. Yes. We use 8 x downsampling for ResNet-101 and it does not have the message passing module of SCNN.
> 2. Softmax is performed in the spatial dimension, i.e., HxW.



> 1. Yes. We use 8 x downsampling for ResNet-101 and it does not have the message passing module of SCNN.
> 2. Softmax is performed in the spatial dimension, i.e., HxW.

谢谢你的回答
1.对于第一个回答，resnet101中是没有加入空洞卷积是吗？
2.对应第2个回答，我想再确定一下，您的意思是，原来卷积的输出特征图（C, H, W）经过在channel上求和变成（H, W)的tensor， 然后，在这个（H, W）上使用spatial  softmax之后， 这个（H, W）中所有像素值之和等于1.是这样的吗？
再次感谢！

**Comment by cardwing** [comment]
1. Dilated convolution is used for ResNet-101, which is the same as the deeplab version. You can refer to the [SCNN paper](https://pdfs.semanticscholar.org/28f5/3ec7732299fa946ed3fc27bf691a6ab5c60c.pdf?_ga=2.120853576.1999152046.1567408404-1411737913.1538625169) for more details.
2. Yes.

**Comment by harryhan618** [comment]
Hi, may I ask a question about `SpatialSoftMax` in lua torch? What's difference between `cudnn.SpatialSoftMax` and `nn.SpatialSoftMax`?
In `ENet-Label-Torch`,  in exist branch, there is a `cudnn.SpatialSoftMax` before `Linear` layer. I suppose it is done along channel dimension. 
And in SAD module, there is a `nn.SpatialSoftMax`. I suppose it is done in spatial dimension, i.e., H*W.

I wanna know whether what I understand is correct.

**Comment by cardwing** [comment]
Yes. And I think `cudnn.SpatialSoftMax` should run much faster than `nn.SpatialSoftMax`.

**Comment by pgr2015** [comment]
@cardwing thank you very much for your great work. I am very interested in SAD. Since I am not familar with lua, so could you please point out the SpatialSoftMax funtion in your code? Thanks in advance!

**Comment by cardwing** [comment]
It is in the trained model. You can check it when the model architecture is printed.

**Comment by pgr2015** [comment]
@cardwing thank you for your reply. 
https://github.com/cardwing/Codes-for-Lane-Detection/blob/master/ENet-TuSimple-Torch/models/init.lua#L38
This is source code, if I understand correctly, softmax will be added when training process has gone half way.
